pkg/reconcile/runner.go

```diff
@@ -0,0 +1,158 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type Reconciler interface {
+	// reconcile an object
+	// requeue the object if returning an error, or a non-zero "requeue-after" duration
+	Reconcile(object ezkube.Object) (time.Duration, error)
+}
+
+type FinalizingReconciler interface {
+	Reconciler
+
+	// name of the finalizer used by this handler.
+	// finalizer names should be unique for a single task
+	FinalizerName() string
+
+	// finalize the object before it is deleted.
+	// Watchers created with a finalizing handler will a
+	Finalize(object ezkube.Object) error
+}
+
+// a Reconcile Loop runs resource reconcilers until the context gets cancelled
+type Loop interface {
+	RunReconciler(ctx context.Context, reconciler Reconciler, predicates ...predicate.Predicate) error
+}
+
+type runner struct {
+	name     string
+	mgr      manager.Manager
+	resource ezkube.Object
+}
+
+func NewLoop(name string, mgr manager.Manager, resource ezkube.Object) *runner {
+	return &runner{name: name, mgr: mgr, resource: resource}
+}
+
+type runnerReconciler struct {
+	ctx        context.Context
+	mgr        manager.Manager
+	resource   ezkube.Object
+	logger     logr.Logger
+	reconciler Reconciler
+}
+
+func (ec *runner) RunReconciler(ctx context.Context, reconciler Reconciler, predicates ...predicate.Predicate) error {
+	gvk, err := apiutil.GVKForObject(ec.resource, ec.mgr.GetScheme())
+	if err != nil {
+		return err
+	}
+	rec := &runnerReconciler{
+		logger:     log.Log.WithName("event-controller").WithValues("kind", gvk).WithName(ec.name),
+		mgr:        ec.mgr,
+		resource:   ec.resource,
+		reconciler: reconciler,
+	}
+
+	ctl, err := controller.New(ec.name, ec.mgr, controller.Options{
+		Reconciler: rec,
+	})
+	if err != nil {
+		return err
+	}
+
+	// send us watch events
+	if err := ctl.Watch(&source.Kind{Type: ec.resource}, &handler.EnqueueRequestForObject{}, predicates...); err != nil {
+		return err
+	}
+
+	rec.logger.V(1).Info("waiting for cache sync...")
+	if synced := ec.mgr.GetCache().WaitForCacheSync(ctx.Done()); !synced {
+		return errors.Errorf("waiting for cache sync failed")
+	}
+
+	return nil
+}
+
+func (ec *runnerReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := ec.logger.WithValues("event", request)
+	logger.V(2).Info("handling event", "event", request)
+
+	// get the object from our cache
+	restClient := ezkube.NewRestClient(ec.mgr)
+
+	obj := ec.resource.DeepCopyObject().(ezkube.Object)
+	obj.SetName(request.Name)
+	obj.SetNamespace(request.Namespace)
+	if err := restClient.Get(ec.ctx, obj); err != nil {
+		logger.Error(err, "unable to fetch %T %v", obj, request)
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return reconcile.Result{}, client.IgnoreNotFound(err)
```

Comment:
Can we log the object cannot be found, could be a confusing error hole otherwise. Granted this should never happen